builtin/k8s/releaser.go

```diff
@@ -339,10 +353,367 @@ func (r *Releaser) resourceServiceDestroy(
+	}
+
+	if r.config.IngressConfig.ClassName != "http" {
+		return status.Error(codes.FailedPrecondition, "An ingress stanza must be "+
+			"of type \"http\".")
+	}
+
+	// Prepare our namespace and override if set.
+	ns := csinfo.Namespace
+	if r.config.Namespace != "" {
+		ns = r.config.Namespace
+	}
+
+	step := sg.Add("Preparing ingress resource...")
+	defer func() { step.Abort() }() // Defer in func in case more steps are added to this func in the future
+
+	clientSet := csinfo.Clientset
+	serviceclient := clientSet.CoreV1().Services(ns)
+	ingressClient := clientSet.NetworkingV1().Ingresses(ns)
+
+	// Determine if we have a deployment that we manage already
+	serviceBackend, err := serviceclient.Get(ctx, serviceState.Name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		// We expect resourceServiceCreate to have created a Service prior to
+		// creating an ingress resource. Otherwise there is no service backend
+		// the ingress resource can refer to
+		return status.Errorf(codes.FailedPrecondition, "A service must exist before "+
+			"an ingress resource can be created: %s", err)
+	}
+	if err != nil {
+		return err
+	}
+
+	var ingressResource *networkingv1.Ingress
+	create := false
+	ingressResource, err = ingressClient.Get(ctx, serviceState.Name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		// We haven't created an ingress resource yet...
+		log.Debug("no ingress resource found, will create one")
+
+		err = nil
+		create = true
+
+		// basic ingress resource
+		ingressResource = &networkingv1.Ingress{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "Ingress",
+			},
+
+			ObjectMeta: metav1.ObjectMeta{
+				Name: result.ServiceName,
+			},
+		}
+	}
+	if err != nil {
+		return err
+	}
+
+	// Set the ingress resource state name to match the service name
+	state.Name = result.ServiceName
+
+	var ingressTls networkingv1.IngressTLS
+	if r.config.IngressConfig.TlsConfig != nil {
+		ingressTls = networkingv1.IngressTLS{
+			Hosts:      r.config.IngressConfig.TlsConfig.Hosts,
+			SecretName: r.config.IngressConfig.TlsConfig.SecretName,
+		}
+	}
+
+	// Apply any annotations to the ingress resource
+	if r.config.IngressConfig.Annotations != nil {
+		ingressResource = &networkingv1.Ingress{
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: r.config.IngressConfig.Annotations,
+			},
+		}
+	}
+
+	// Define the ingress resources backend service it should route traffic to
+	ingressBackend := networkingv1.IngressBackend{
+		Service: &networkingv1.IngressServiceBackend{
+			Name: serviceState.Name,
+			Port: networkingv1.ServiceBackendPort{
+				Name:   serviceBackend.Spec.Ports[0].Name,
+				Number: serviceBackend.Spec.Ports[0].Port,
+			},
+		},
+	}
+
+	// NOTE, the k8s go api doesn't seem to like being set directly from the
+	// string config in our releaser, so we manually set it with this switch
+	var pathType networkingv1.PathType
+	switch r.config.IngressConfig.PathType {
+	case "Exact":
+		pathType = "Exact"
```

Comment:
Are there constants in the networkingv1 package we can use instead of strings here? Since this appears to be an enum I'd rather use something stronger typed than a string.

Reply:
I believe the values were constants yeah! I'll give that a shot instead

Reply:
Yeah that worked! 👍🏻